node/src/accountant/scanners/pending_payable_scanner/utils.rs

```diff
@@ -1,191 +1,1184 @@
+    pub tx_failures: Vec<PresortedTxFailure>,
+    pub tx_receipt_rpc_failures: Vec<FailedValidationByTable>,
+}
+
+impl DetectedFailures {
+    fn requires_retry(&self) -> Option<Retry> {
+        if self.tx_failures.is_empty() && self.tx_receipt_rpc_failures.is_empty() {
+            None
+        } else if !self.tx_failures.is_empty() {
+            Some(Retry::RetryPayments)
+        } else {
+            Some(Retry::RetryTxStatusCheckOnly)
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum PresortedTxFailure {
+    NewEntry(FailedTx),
+    RecheckCompleted(TxHash),
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum FailedValidationByTable {
+    SentPayable(FailedValidation<TxStatus>),
+    FailedPayable(FailedValidation<FailureStatus>),
+}
+
+impl From<(TxReceiptError, TxStatus)> for FailedValidationByTable {
+    fn from((tx_receipt_error, current_status): (TxReceiptError, TxStatus)) -> Self {
+        match tx_receipt_error.tx_hash {
+            TxHashByTable::SentPayable(tx_hash) => Self::SentPayable(FailedValidation::new(
+                tx_hash,
+                BlockchainErrorKind::AppRpc(tx_receipt_error.err.into()),
+                current_status,
+            )),
+
+            TxHashByTable::FailedPayable(tx_hash) => {
+                unreachable!(
+                    "Mismatch in the type of tx record (failed tx) and status type (TxStatus) for {:?}", tx_hash
+                )
+            }
+        }
+    }
+}
+
+impl From<(TxReceiptError, FailureStatus)> for FailedValidationByTable {
+    fn from((tx_receipt_error, current_status): (TxReceiptError, FailureStatus)) -> Self {
+        match tx_receipt_error.tx_hash {
+            TxHashByTable::FailedPayable(tx_hash) => Self::FailedPayable(FailedValidation::new(
+                tx_hash,
+                BlockchainErrorKind::AppRpc(tx_receipt_error.err.into()),
+                current_status,
+            )),
+            TxHashByTable::SentPayable(tx_hash) => {
+                unreachable!(
+                    "Mismatch in the type of tx record (sent tx) and status type (FailureStatus) for {:?}",tx_hash
+                )
+            }
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct FailedValidation<RecordStatus> {
+    pub tx_hash: TxHash,
+    pub validation_failure: BlockchainErrorKind,
+    pub current_status: RecordStatus,
+}
+
+impl<RecordStatus> FailedValidation<RecordStatus>
+where
+    RecordStatus: UpdatableValidationStatus,
+{
+    pub fn new(
+        tx_hash: TxHash,
+        validation_failure: BlockchainErrorKind,
+        current_status: RecordStatus,
+    ) -> Self {
+        Self {
+            tx_hash,
+            validation_failure,
+            current_status,
+        }
+    }
+
+    pub fn new_status(&self, clock: &dyn ValidationFailureClock) -> Option<RecordStatus> {
+        self.current_status
+            .update_after_failure(self.validation_failure.clone(), clock)
```

Comment:
I think `BlockchainErrorKind` should implement `Copy`, and we shouldn't be using clone here.